### ISSUE TITLE
RFC: Streaming Conformer with LSTM Recurrence (`muaalem_streaming_lstm`)

### ISSUE THREAD

**Issue by obadx** [obadx/quran-muaalem#31]
## High-Level Goal

Create (or refine) a streaming Conformer module that uses a **small local context** processed by the Conformer encoder, with an **LSTM recurrent layer** to carry history across chunks. No KV caching, no long attention context. This is Architecture B from Issue #10.

Each chunk: **25 frames (500ms)** with **5 frames (100ms) lookahead**. The Conformer processes only a **local window of 5 frames** of actual context, and relies entirely on LSTM recurrence to propagate information across the utterance.

## Architecture Diagram (from Issue #10)

![Architecture](https://private-user-images.githubusercontent.com/16362655/576987490-ef66c3b9-6d33-4ada-9c32-fc0ca3ec7775.png)

*Source: https://github.com/obadx/quran-muaalem/issues/10*

## Module

```
conformer-python/src/conformer_python/muaalem_streaming_lstm/
├── __init__.py
├── configuration_multi_level_ctc_streaming_lstm.py
├── modeling_multi_level_ctc_streaming_lstm.py
├── multi_level_tokenizer.py
└── vocab.py
```

## Bold Lines

1. **Small local context** — The Conformer encoder processes only a short window. Unlike `muaalem_streaming_kv_cache` which keeps 250 frames of attention context, this model uses **5 frames** of actual encoder attention context. Total per-chunk input: 25 + 5 = 30 frames.

2. **LSTM for long-range memory** — After the Conformer encoder processes a chunk, its output is fed through an LSTM. The LSTM's `(h_n, c_n)` is carried across chunks providing indefinite history without storing past key/value pairs.

3. **Stateful forward** — Forward accepts and returns `past_lstm_state` / `present_lstm_state`, analogous to KV cache. Caller feeds chunks sequentially.

4. **No KV caching** — Attention is limited to a small local window; all temporal history flows through LSTM recurrence.

5. **Causal convolution** — Existing `Wav2Vec2BertConvolutionModule` reused as-is (already causal).

6. **Multi-level CTC heads** — Same pattern as offline. Logits concatenated along time axis during training.

## Frame arithmetic

| Parameter | Value | Description |
|---|---|---|
| `chunk_size` | 25 | Frames per step (500ms at 50Hz) |
| `lookahead_frames` | 5 | Future frames (100ms) |
| `encoder_context` | 5 | Attention/conv context window per frame |
| `total_chunk_input` | 30 | `chunk_size + lookahead_frames` |
| Feature rate | 50 Hz | 20ms per frame |

## Configuration (`Wav2Vec2BertForMultilevelCTCStreamingLSTMConfig`)

Same as offline config, plus:

- `chunk_size` (default: 25)
- `lookahead_frames` (default: 5)
- `encoder_context` (default: 5)
- `lstm_hidden_size` (default: `config.hidden_size`)
- `lstm_output_size` (default: `config.lstm_output_size` make it default to hidden_size /2)
- `lstm_num_layers` (default: 1)
- `lstm_dropout` (default: config.dropout)

## Model architecture

```
Wav2Vec2BertForMultilevelCTCStreamingLSTM
├── Wav2Vec2BertModel
│   ├── FeatureProjection
│   └── Wav2Vec2BertEncoder (local context only)
│       └── Wav2Vec2BertEncoderLayer × N
│           ├── FFN1
│           ├── SelfAttention (local window)
│           ├── ConvModule (causal)
│           └── FFN2
├── LSTM (input_size=hidden_size, hidden_size=lstm_hidden_size, num_layers=lstm_num_layers)
├── Dropout
└── Multi-level CTC heads (linear projections)
```

## Forward pass contract

```python
def forward(
    self,
    input_features: Tensor,              # (batch, chunk_size + lookahead, feat_dim)
    attention_mask: Optional[Tensor],
    past_lstm_state: Optional[tuple]     # (h_n, c_n) from previous chunk
) -> tuple[Tensor, tuple]:
    # returns (logits, updated_lstm_state)
```

LSTM state shapes: `(num_layers, batch, lstm_hidden_size)` each.

## Training

- Full utterances sliced into overlapping windows of 30 frames (25 + 5 lookahead)
- Step: 25 frames between windows
- Encoder processes each window independently
- LSTM state carried sequentially across windows; reset between utterances
- CTC loss over concatenated logits with blank insertion (per Issue #10)

## Development Place
The development will take place in the [conformer-rs](https://github.com/obadx/conformer-rs) repo not here.

## After Designing

After developing the architecture run it on-device for testing

## References

- Issue #10: https://github.com/obadx/quran-muaalem/issues/10
- Issue #29: https://github.com/obadx/quran-muaalem/issues/29
- Existing `muaalem_streaming_lstm/` (reference implementation)
- Nvidia FastConformer: https://arxiv.org/pdf/2305.05084
- Nemo Streaming: [here](https://docs.nvidia.com/nemo-framework/user-guide/latest/nemotoolkit/asr/intro.html)
